handlers.go

```diff
@@ -3526,3 +3528,64 @@ func (s *Server) handleGetPublicNodeInfo(c echo.Context) error {
 		PrimaryAddress: s.FilClient.ClientAddr,
 	})
 }
+
+type retrievalCandidate struct {
+	maddr   address.Address
+	rootCid cid.Cid
+}
+
+func (s *Server) handleGetRetrievalCandidates(c echo.Context) error {
+	// Read the cid from the client request
+	cid, err := cid.Decode(c.Param("cid"))
+	if err != nil {
+		return c.String(http.StatusBadRequest, "Invalid cid")
+	}
+
+	// Get the object with the requested cid
+	var object Object
+	if err := s.DB.Take(&object, "cid = ?", cid).Error; err != nil {
+		return err
+	}
+
+	// Get all the refs for that object
+	var objRefs []ObjRef
+	if err := s.DB.Find(&objRefs, "object = ?", object.ID).Error; err != nil {
+		return err
+	}
+
+	// Turn object refs' content IDs into their own list
+	var contentIDs []uint
+	for _, objRef := range objRefs {
+		contentIDs = append(contentIDs, objRef.Content)
+	}
+
+	// Get the deals for each of those object refs' contents
+	var candidateInfos []struct {
+		miner string
+		cid   util.DbCID
+	}
+	if err := s.DB.
+		Table("content_deals").
+		Where("content IN ? AND NOT failed ", contentIDs).
+		Joins("JOIN contents ON content_deals.content = content.id").
```

Comment:
I think you want to say "left join" here?

Also, I think we can do this whole query in one go:
```sql
select miner, cid from content_deals left 
  join contents on content_deals.content = contents.id
  where contents.id in (
    select content from obj_refs
      where object in (select id from objects where cid = ?)
  );
```

Reply:
(that is, we can do away with lines 3544 through 3565)

Reply:
Aside from me playing sql golf though, Your solution here seems correct (potentially minus the 'left' qualifier on the join)

Reply:
i've attempted to implement this, although let me know if it is going to do something crazy